File: leetcode/medium/853_car_fleet/car_fleet.py
from typing import List


class Solution:
    def carFleet(self, target: int, position: List[int], speed: List[int]) -> int:
        cars = list(zip(position, speed))
        cars = sorted(cars, key=lambda x: x[0])
        time_to_finish = list(map(lambda x: (target - x[0]) / x[1], cars))
        current_min_time = time_to_finish[-1]
        for i in reversed(range(len(time_to_finish))):
            if time_to_finish[i] < current_min_time:
                time_to_finish[i] = current_min_time
            else:
                current_min_time = time_to_finish[i]
        return len(set(time_to_finish))


assert Solution().carFleet(12, [10, 8, 0, 5, 3], [2, 4, 1, 1, 3]) == 3
assert Solution().carFleet(10, [8, 3, 7, 4, 6, 5], [4, 4, 4, 4, 4, 4]) == 6
assert Solution().carFleet(100, [0, 2, 4], [4, 2, 1]) == 1

File: leetcode/medium/853_car_fleet/car_fleet.cpp
#include <algorithm>
#include <cstdio>
#include <unordered_set>
#include <utility>
#include <vector>

class Solution {
public:
  int carFleet(int target, std::vector<int> &position,
               std::vector<int> &speed) {
    std::vector<std::pair<int, int>> cars;
    for (int i = 0; i < position.size(); i++) {
      cars.emplace_back(position[i], speed[i]);
    }
    std::sort(cars.begin(), cars.end(),
              [](std::pair<int, int> a, std::pair<int, int> b) {
                return a.first < b.first;
              });
    std::vector<float> time_to_finish;
    for (auto car : cars) {
      auto ttf = ((float)target - car.first) / car.second;
      time_to_finish.push_back(((float)target - car.first) / car.second);
    }
    float current_min_time = time_to_finish[time_to_finish.size() - 1];
    for (int i = time_to_finish.size() - 1; i >= 0; i--) {
      if (time_to_finish[i] < current_min_time) {
        time_to_finish[i] = current_min_time;
      } else {
        current_min_time = time_to_finish[i];
      }
    }
    return std::unordered_set<float>(time_to_finish.begin(),
                                     time_to_finish.end())
        .size();
  }
};
